### DISCUSSION TITLE
Applying External Force in Compressible Channel Flow Simulation

### DISCUSSION THREAD

**Discussion by JhonCordova** [AMReX-Combustion/PeleC#861] [Q&A]
Hi everyone,

I want to apply and external force `f_i` to analize compressible channel flow. As shown in the figure above, `f_i` depends on the viscous stress tensor `τ_12` at the walls `y=H` and `y=-H`.

![imagen](https://github.com/user-attachments/assets/6dd6a690-ee6b-4674-9c42-937c60697b79)

I can extract velocities at the cell closest to the wall using `Farrs[nbx](i, j, k, URHO)` and `Farrs[nbx](i, j, k, RHO)` by setting j to the cell number nearest to the wall. However, I am uncertain about the formulation of the gradients.

Do you have any sugestion on how to obtain this information using the following this: [https://github.com/baperry2/PeleC/blob/54c8253eb53e887d45cf436d26dbb6d549d8de53/Exec/RegTests/EB-Vane/prob.H#L125](url)

**Comment by baperry2** [comment]
This should be possible to do but not trivial. The `problem_modify_ext_sources` function gives you access to the full state multifab, but note that you can't assume any given box of data in the AMR hierarchy will overlap the wall boundary, so you can't directly index into (i,j,k) arbitrarily. 

You'd need to do this in stages: first do an MFiter loop and take the intersection of each box with the boxes of cells along the walls, and compute shear stress in a ParallelFor loop over the intersection boxes. Then use parallel communication to get the average shear stress across all the intersection boxes. Then do another MFIter/ParallelFor over all the data and add the desired source term based on the computed shear stress. This will take a fairly detailed understanding of how the AMReX data structures work, so feel free to ask follow up questions as you try this.
